Move goto_convert files

kroening opened the pull request:
The `goto_convertt` class and associated helpers convert a C parse tree into a set of GOTO functions.  They are specific to C, and hence, should be in the `ansi-c/` directory.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented on jbmc/src/java_bytecode/assignments_from_json.cpp:
```diff
@@ -16,7 +16,7 @@ Author: Diffblue Ltd.
 #include <util/unicode.h>
 
-#include <goto-programs/allocate_objects.h>
+#include <ansi-c/allocate_objects.h>
```
Is this one really C specific? It appears to be used by JBMC, and perhaps more so than anywhere else?

kroening replied:
I'd say it tries to do too much. Object construction is clearly language specific, and this should get split up.

tautschnig commented:
I have a couple of concerns:
* The ansi-c folder is starting to become unmanageable. At bare minimum I'd suggest that we have a new subfolder `ansi-c/goto-conversion/` (or some similar name) to hold those files.
* As indicated by my comment, but perhaps applicable in more cases: are all those files really all about goto conversion?
* Is it really correct to assert that only C parse trees can be converted? Isn't it rather the case that we actually have a further intermediate language of `codet` statements?
* Do we perhaps need to become a lot more clinical on where we use `goto_convert`? As this PR documents, we make use of it in a number of places well outside the C front-end. To that effect, I'd like to put up for debate whether we do such cleanup before moving files.

kroening commented:
> * The ansi-c folder is starting to become unmanageable. At bare minimum I'd suggest that we have a new subfolder `ansi-c/goto-conversion/` (or some similar name) to hold those files.

Ok, will do.

> * As indicated by my comment, but perhaps applicable in more cases: are all those files really all about goto conversion?

No, there's also object construction, but that appears to be intertwined with the conversion. As said in the comment about it, it's also language specific, and hence should move out of the goto-programs folder.

> * Is it really correct to assert that only C parse trees can be converted? Isn't it rather the case that we actually have a further intermediate language of `codet` statements?

This is getting reduced, and I'd like to get rid of it altogether.

> * Do we perhaps need to become a lot more clinical on where we use `goto_convert`? As this PR documents, we make use of it in a number of places well outside the C front-end. To that effect, I'd like to put up for debate whether we do such cleanup before moving files.

We realistically won't do the cleanup. For symtab2gb, this lives outside of the codebase. For jbmc, we don't have the person to do it.

tautschnig commented:
> > * The ansi-c folder is starting to become unmanageable. At bare minimum I'd suggest that we have a new subfolder `ansi-c/goto-conversion/` (or some similar name) to hold those files.
> 
> Ok, will do.
[...]

Happy to approve once this is done: I think this goes a long way towards making all of my other concerns easy to track (and, if resources ever permit: fix).

kroening commented:
Now with ansi-c/goto-conversion subdirectory.

kroening committed:
move goto_convert from goto-programs/ to ansi-c/

The goto_convertt class and associated helpers convert a C parse tree into a
set of GOTO functions.  They are specific to C, and hence, should be in the
ansi-c/ directory.

kroening committed:
clang-format the moved files

peterschrammel commented on src/linking/module_dependencies.txt:
```diff
@@ -1,3 +1,4 @@
+ansi-c
```
❓ Isn't linking a pure GOTO thing? Why is there a dependency on ansi-c now? This seems to indicate that either some functionality has been moved that should not be moved or that some functionality is used in other modules that shouldn't be used there.

kroening replied:
The culprit for this is `recreate_initialize_function`. I'd like to kill it, and one benefit would be to remove this coupling.

tautschnig replied:
Being the one who introduced this function in the first place: can you share some ideas on how to replace it?

kroening replied:
As mentioned out-of-band, simply create the goto instructions directly. This should be done separately.
